File: 543-DiameterOfBinaryTree/543-DiameterOfBinaryTree.py
# Last updated: 6/12/2025, 8:29:51 AM
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.max_height = 0
        self.depth_of_tree(root)
        return self.max_height
    def depth_of_tree(self, new_root):
            if new_root == None:
                return 0
            else:
                l = self.depth_of_tree(new_root.left)
                r = self.depth_of_tree(new_root.right)
                curr_height = l+r
                self.max_height = max(self.max_height, curr_height)
                print(self.max_height, curr_height)
                return max(l,r) + 1

File: 543-DiameterOfBinaryTree/543-DiameterOfBinaryTree.java
// Last updated: 6/12/2025, 8:29:54 AM
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public static int count;
    public int diameterOfBinaryTree(TreeNode root) {
        count =0;
        longest(root);
        return count;
    }
    
    public static int longest(TreeNode root){
        if(root == null){return 0;}
        int l = longest(root.left);
        int r = longest(root.right);
        
        count = Math.max(l+r,count);
        return Math.max(l,r)+1;
    }
}